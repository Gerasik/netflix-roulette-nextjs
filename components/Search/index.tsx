import React, { FunctionComponent } from 'react';
import classNames from 'classnames';

import { Action } from 'containers/Search/models';
import styles from './style.module.scss';
import { SearchProps } from './models';

const Search: FunctionComponent<SearchProps> = ({
  searchData,
  changeSearchString,
  changeSearchBy,
  setData,
}) => {
  const { searchString, searchBy, sortBy } = searchData;

  function handleClickSearch(): void {
    fetch(
      `https://reactjs-cdp.herokuapp.com/movies?sortBy=${sortBy}&sortOrder=asc&search=${searchString}&searchBy=${searchBy}&limit=6`
    )
      .then(response => response.json())
      .then(response => {
        setData(response);
      });
  }

  return (
    <div className={styles.root}>
      <p className={styles.label}>find your movie</p>
      <div className={styles['search-form']}>
        <input
          type="text"
          placeholder="Search"
          className={styles['search-text']}
          onChange={(event): Action.ChangeSearchString => changeSearchString(event.target.value)}
          value={searchString}
        />
        <button type="button" className={styles['search-button']} onClick={handleClickSearch}>
          search
        </button>
      </div>
      <div className={styles['search-by']}>
        <span>search by</span>
        <button
          type="button"
          className={classNames(styles['by-title'], { [styles.active]: searchBy === 'title' })}
          onClick={(): Action.ChangeSearchBy => changeSearchBy('title')}
        >
          title
        </button>
        <button
          type="button"
          className={classNames(styles['by-genre'], { [styles.active]: searchBy === 'genres' })}
          onClick={(): Action.ChangeSearchBy => changeSearchBy('genres')}
        >
          genre
        </button>
      </div>
    </div>
  );
};

export default Search;
