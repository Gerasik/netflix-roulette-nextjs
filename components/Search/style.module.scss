@import '../../styles/colors.scss';

$borderRadiusSize: 0.3em;

.root {
  color: $fontColor;
  width: 100%;
  button {
    border: none;
    color: $fontColor;
    text-transform: uppercase;
  }
  .label {
    font-size: 2em;
    margin: 0.4em 0;
  }
  .label,
  .search-by {
    font-weight: 100;
    text-transform: uppercase;
  }
  .search-form {
    display: flex;
    justify-content: stretch;
    .search-text,
    .search-button {
      font-size: 1.4em;
      border-radius: $borderRadiusSize;
    }
    .search-text {
      color: $fontColor;
      width: 100%;
      margin-right: 0.5em;
      border: none;
      background-color: rgba($color: $primaryColor, $alpha: 0.8);
      padding: 0.5em 1em;
    }
    .search-button {
      padding: 0.5em 2em;
      background-color: $secondaryFontColor;
      cursor: pointer;
    }
  }
  .search-by {
    font-size: 0.9em;
    margin: 1.3em 0;
    span {
      padding-right: 1em;
    }
    .by-title {
      border-radius: $borderRadiusSize 0em 0em $borderRadiusSize;
    }
    .by-genre {
      border-radius: 0 $borderRadiusSize $borderRadiusSize 0;
    }
    .by-genre,
    .by-title {
      font-size: 0.9em;
      padding: 0.5em 2em;
      background-color: rgba($color: $tertiaryColor, $alpha: 0.6);
      cursor: pointer;
      &:hover {
        background-color: $secondaryFontColor;
      }
    }
    .active {
      background-color: $secondaryFontColor;
    }
  }
}
