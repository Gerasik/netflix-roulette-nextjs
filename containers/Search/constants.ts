const ActionType = {
  CHANGE_SEARCH_STRING: 'CHANGE-SEARCH-STRING',
  CHANGE_SEARCH_BY: 'CHANGE-SEARCH-BY',
  CHANGE_SORT_BY: 'CHANGE-SORT-BY',
};

export default ActionType;
