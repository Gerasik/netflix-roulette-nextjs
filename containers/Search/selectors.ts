import { createSelector } from 'reselect';

import { State } from './models';

const searchInfo = (state): State => {
  return state.get('searchReducer') as State;
};

export const searchData = createSelector(searchInfo, searchInfo => ({
  searchString: searchInfo.get('searchString'),
  searchBy: searchInfo.get('searchBy'),
  sortBy: searchInfo.get('sortBy'),
}));
