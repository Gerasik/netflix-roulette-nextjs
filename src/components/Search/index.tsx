import React, { FunctionComponent, useEffect } from 'react';
import classNames from 'classnames';
import { debounce } from 'lodash';
import { useRouter } from 'next/router';

import { Action } from 'containers/Search/models';
import styles from './style.module.scss';
import { SearchProps } from './models';

const Search: FunctionComponent<SearchProps> = ({
  searchData,
  changeSearchString,
  changeSearchBy,
  setStartData,
}) => {
  const { searchBy, searchString } = searchData;
  const router = useRouter();
  const { query } = router.query;

  useEffect(() => {
    changeSearchString(query as string);
    setStartData();
  }, [query]);

  function handleClickSearch(): void {
    setStartData();
  }

  function handleChangeSearch(event): void {
    const searchString = event.target.value;
    const href =
      window.location.href.slice(-6) === 'search' ? `search/${searchString}` : searchString;

    window.history.pushState(null, '', href || '/search');
    setStartData();
  }

  function debounceEvent(...arg) {
    const fn = arg[0];
    const debounceEvent = debounce(fn, 500);
    return e => {
      changeSearchString(e.target.value);
      e.persist();
      return debounceEvent(e);
    };
  }

  return (
    <div className={styles.root}>
      <p className={styles.label}>find your movie</p>
      <div className={styles['search-form']}>
        <input
          value={searchString}
          type="text"
          placeholder="Search"
          className={styles['search-text']}
          onChange={debounceEvent(handleChangeSearch)}
        />
        <button type="button" className={styles['search-button']} onClick={handleClickSearch}>
          search
        </button>
      </div>
      <div className={styles['search-by']}>
        <span>search by</span>
        <button
          type="button"
          className={classNames(styles['by-title'], { [styles.active]: searchBy === 'title' })}
          onClick={(): Action.ChangeSearchBy => changeSearchBy('title')}
        >
          title
        </button>
        <button
          type="button"
          className={classNames(styles['by-genre'], { [styles.active]: searchBy === 'genres' })}
          onClick={(): Action.ChangeSearchBy => changeSearchBy('genres')}
        >
          genre
        </button>
      </div>
    </div>
  );
};

export default Search;
