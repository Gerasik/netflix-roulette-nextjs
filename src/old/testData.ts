/* eslint-disable @typescript-eslint/camelcase */
import { fromJS } from 'immutable';

import { Movie, MoviesResponse, MoviesData, MovieMap } from '../../models';
import { State } from '../../containers/Body/models';

export const testMovie1: Movie | MovieMap = {
  id: 1,
  title: 'first movie',
  tagline: 'test',
  vote_average: 55,
  vote_count: 66,
  release_date: '2020-05-01',
  poster_path: '/movie_url/image.jpg',
  overview: 'ove',
  budget: 54545,
  revenue: 5454,
  runtime: 5454,
  genres: ['drama', 'sitcom', 'drama'],
};

export const testMovie2: Movie | MovieMap = {
  id: 2,
  title: 'second movie',
  tagline: 'test',
  vote_average: 55,
  vote_count: 66,
  release_date: '2020-05-01',
  poster_path: '/movie_url/image2.jpg',
  overview: 'ove',
  budget: 54545,
  revenue: 5454,
  runtime: 5454,
  genres: ['drama'],
};

export const testMoviesData: MoviesData = fromJS([testMovie1, testMovie2]);

export const testMovieResponse: MoviesResponse = fromJS({
  data: testMoviesData,
  total: 5,
  offset: 5,
  limit: 5,
});

export const testBodyReducerData = {
  moviesResponse: {},
};
export const testSearchReducerData = {
  searchString: '',
  searchBy: 'genres',
  sortBy: 'vote_average',
};

export const testInitialState: State = fromJS({
  bodyReducer: testBodyReducerData,
  searchReducer: testSearchReducerData,
});

export const testEmptyInitialState: State = fromJS({
  bodyReducer: {
    data: [],
    total: 0,
    offset: 5,
    limit: 5,
  },
  searchReducer: testSearchReducerData,
});

export const testNextState: State = fromJS({
  moviesResponse: testMovieResponse,
});
