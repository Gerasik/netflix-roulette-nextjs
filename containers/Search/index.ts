import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';

import Search from '../../components/Search';
import { setData } from '../../containers/Body/actions';
import { changeSearchString, changeSearchBy } from './actions';
import { searchData } from './selectors';

const mapStateToProps = createStructuredSelector({
  searchData,
});

const mapDispatchToProps = {
  changeSearchString,
  changeSearchBy,
  setData,
};

export default connect(mapStateToProps, mapDispatchToProps)(Search);
