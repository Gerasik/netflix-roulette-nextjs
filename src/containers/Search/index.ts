import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';

import Search from 'components/Search';
import { setStartData } from 'containers/Body/actions';
import { changeSearchString, changeSearchBy } from './actions';
import { searchData } from './selectors';

const mapStateToProps = createStructuredSelector({
  searchData,
});

const mapDispatchToProps = {
  changeSearchString,
  changeSearchBy,
  setStartData,
};

export default connect(mapStateToProps, mapDispatchToProps)(Search);
