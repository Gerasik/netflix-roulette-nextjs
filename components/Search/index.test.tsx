import React from 'react';
import { shallow } from 'enzyme';

import { testSearchReducerData, testMovieResponse } from 'testData';
import Search from '.';

const searchComponent = (
  <Search
    searchData={testSearchReducerData}
    changeSearchString={() => ({ type: 'string', payload: 'testMovieResponse' })}
    changeSearchBy={() => ({ type: 'string', payload: 'testMovieResponse' })}
    setData={() => ({ type: 'string', payload: testMovieResponse })}
  />
);

it('render correctly date component', () => {
  const wrapper = shallow(searchComponent);
  wrapper.find('.search-button').simulate('click');
  wrapper.find('.by-title').simulate('click');
  wrapper.find('.by-genre').simulate('click');
  wrapper.find('.search-text').simulate('change', { target: { value: 'string' } });
  expect(wrapper).toMatchSnapshot();
});

describe('<Search />', () => {
  it('renders an `footer`', () => {
    const wrapper = shallow(searchComponent);
    expect(wrapper.find('input').prop('placeholder')).toBe('Search');
  });
});
