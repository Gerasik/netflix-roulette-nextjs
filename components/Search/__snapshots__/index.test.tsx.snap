// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly date component 1`] = `
<Fragment>
  <p
    className="label"
  >
    find your movie
  </p>
  <div
    className="search-form"
  >
    <input
      className="search-text"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <button
      className="search-button"
      onClick={[Function]}
      type="button"
    >
      search
    </button>
  </div>
  <div
    className="search-by"
  >
    <span>
      search by
    </span>
    <button
      className="by-title"
      onClick={[Function]}
      type="button"
    >
      title
    </button>
    <button
      className="by-genre active"
      onClick={[Function]}
      type="button"
    >
      genre
    </button>
  </div>
</Fragment>
`;
